### ISSUE TITLE
Code to evaluate your CLAP model

### ISSUE THREAD

**Issue by GeneveyC** [yyzpiero/RL4RedTeam#2]
Hi,

I tried to get the performance of your model in the evaluation phase. I want to get the performance of your model in the same scenario as the training. I re-used your code and implemented something for the evaluation phase.

```
        agent2 = torch.load("./model_clap/clap_train_on_%s.pt" % args.env_train)
        agent2.eval()

        # Evaluation step
        envs_test = DummyVecEnv([make_env(args.env_test, args.seed, 0)])
        envs_test = VecNormalize(envs_test, norm_obs=True, norm_reward=True)

        step = 0
        next_done = torch.zeros(1).to(device)
        next_obs = torch.Tensor(envs_test.reset()).to(device)
        next_lstm_state = (
            torch.zeros(agent2.lstm.num_layers, 1, agent2.lstm.hidden_size).to(device),
            torch.zeros(agent2.lstm.num_layers, 1, agent2.lstm.hidden_size).to(device),
        )

        if args.coverage:
            converage_hist = np.zeros((1, envs_test.action_space.n))
        else:
            converage_hist = None

        while next_done[0] == False:

            if converage_hist is None:
                action, logprob, _, value, next_lstm_state = agent2.get_action_and_value(next_obs, next_lstm_state, next_done)
            else:
                coverage_hist_norm = matrix_norm(converage_hist)
                coverage_hist_norm = torch.FloatTensor(coverage_hist_norm).to(device).detach()
                action, logprob, _, value, next_lstm_state = agent2.get_action_and_value(next_obs, next_lstm_state, next_done, coverage_hist=coverage_hist_norm)
                converage_hist[range(1), action.cpu()] += 1
         
            next_obs, reward, done, info = envs_test.step(action.cpu().numpy())

            next_obs, next_done = torch.Tensor(next_obs).to(device), torch.Tensor(done).to(device)

            if done[0] == True:
                print("===Done===")
                print(info[0]["episode"]["r"], info[0]["episode"]["l"])
                break
```

After the training of your CLAP model, I used this code to evaluate the same scenario (as the training phase). However, the performance of the CLAP model decreases a lot in the evaluation phase. For example, in the "Tiny-hard" scenario I obtained an episodic return between 187 and 192 (at the end of the training, with total_timesteps=1000000) But in the evaluation phase, I obtained an episodic return equal to 7. I observe the same effect for the "Tiny" and "Tiny-small" scenarios.

1 - Can you confirm that the code I used is correct?
2 - Have you ever tried to evaluate your model in the evaluation phase? If so, did you notice the same effect as I did?

Best regards

**Comment by yyzpiero** [comment]
Hi,

Although it pains me to admit, I faced the same issue as you did and observed poor results in terms of episodic rewards. After delving into the matter, I discovered that the poor evaluation performance was due to the use of vectorized environments, specifically the Vectorization Normalization Wrapper. It's essential to record and apply observation statistics during the evaluation process, which was missing in this case.

This can be done in two possible ways:
1) use the Synchronize Wrapper from OpenAI baselines to synchronize those statistic
2) just  copy the statistics from the training environments to the evaluation environments

I hope this solution helps resolve the problem

Regards

**Comment by GeneveyC** [comment]
Hi, 

Thanks for your answer. Indeed, the problem came from the Vectorization Normalization Wrapper. I modified my code, and now I save the vectorization normalization after the training and load the same for the evaluation, and I obtain the same performance during the training and the evaluation.

Best regard,
